[llvm] add documentation for public interface annotations (LLVM_ABI, etc) (#134710)
llvm/llvm-project: llvm/docs/InterfaceExportAnnotations.rst

LLVM Interface Export Annotations
=================================

With the following build settings, LLVM will be built as a shared library with
hidden default symbol visibility:

::

   LLVM_BUILD_LLVM_DYLIB_VIS=On
   LLVM_BUILD_LLVM_DYLIB=On
   LLVM_LINK_LLVM_DYLIB=On

When building LLVM in this configuration, any symbol intended to be visible to
clients must be explicitly exported.

Exporting a symbol typically requires annotating it with the ``LLVM_ABI`` macro
defined in `compiler.h
<https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/Compiler.h#L152>`__.
There are similar annotations that must be used in a few special cases. This
document describes the common patterns for annotating LLVM symbols for export.

> 

Reviewer: I feel like we should have a section prior to the various cases about the macro itself. In particular, the following bits are useful to cover:

1. naming - `LLVM_ABI` vs `LLVMSupport_ABI` vs `Clang_ABI` vs `libclang_ABI`. The naming is important as it is not a "LLVM project ABI" but rather the "library named 'llvm'".
2. The need for the macro (a cursory explanation of DLL storage)

Functions
---------

Exported function declarations in header files must be annotated with
``LLVM_ABI``.

.. code:: cpp

   #include "llvm/Support/Compiler.h"

   LLVM_ABI void exported_function(int a, int b);

Global Variables
----------------

Exported global variables must be annotated with ``LLVM_ABI`` at their
``extern`` declarations.

.. code:: cpp

   #include "llvm/Support/Compiler.h"

   LLVM_ABI extern int exported_global_variable;

Classes, Structs, and Unions
----------------------------

Classes, structs, and unions can be annotated with ``LLVM_ABI`` at their

> definition, but this option is generally discouraged. Annotating the entire

Reviewer: Do you mean at their declaration? The definition _can_ be marked for export, but not for import and therefore applying `LLVM_ABI` is incorrect for the definition but is proper for definition.

Furthermore, it de-synchronises the declaration from the definition with the attribution.

Author: Yes, I meant "declaration" here. Will correct.

definition exports unnecessary symbols, such as private functions, vtables, and
type information. Instead, ``LLVM_ABI`` should be applied only to public and
protected method declarations without a body in the header, including

> constructors and destructors.

Reviewer: What if I have an inline definition in the body that should be exported? That is, `__declspec(dllexport) inline int f(void) { return 32; }` is a totally well-defined valid construct (as is the mirror'ed `__declspec(dllimport) inline int f(void) { return 32; }`.

I think that there needs to be more clear text here about the application of the attribute and the reasoning.

Author: Do you have a scenario in mind when someone would want to export an inline function? I was going with the assumption that there isn't a case where this is required.

Reviewer: Yes! The following:

```c
__declspec(dllexport) inline int f(void) {
  return 32;
}
```

Is useful for a function that is not referenced within the module (DLL) itself, but is part of its ABI definition. This is then read by the client as:

```c
__declspec(dllimport) inline int f(void) {
  return 32;
}
```

This declaration is particularly interesting with `/Ob1` which would allow calls to `f` to be inlined but would cause `&f` to be referenced with a unique consistent address as the function is homed in the defining module.

Use of PMF as keys is pretty commonplace as a practice.

are exported by default. To build a Windows DLL, however, the situation is more
complex:

- Symbols are not exported from a DLL by default. Symbols must be annotated with
  ``__declspec(dllexport)`` when building the library to be externally visible.

- Symbols imported from a Windows DLL should generally be annotated with
  ``__declspec(dllimport)`` when compiling clients.

- A single Windows DLL can export a maximum of 65,535 symbols.

Annotation Macros
-----------------
The distinct DLL import and export annotations required for Windows DLLs
typically lead developers to define a preprocessor macro for annotating exported
symbols in header public files. The custom macro resolves to the _export_
annotation when building the library and the _import_ annotation when building
the client.

For this purpose, we have defined the `LLVM_ABI` macro in
`llvm/Support/Compiler.h
<https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/Compiler.h#L152>`__.

.. code:: cpp

   #if defined(LLVM_EXPORTS)
   #define LLVM_ABI __declspec(dllexport)
   #else
   #define LLVM_ABI __declspec(dllimport)
   #endif

> Because building LLVM for Windows is less common than ELF and Mach-O platforms,

Reviewer 2: ```suggestion
Building LLVM for Windows is less common than ELF and Mach-O platforms,
```

Reviewer 2: Probably also worth saying what elf, mach-o and coff is in a succinct form.

> Because building LLVM for Windows is less common than ELF and Mach-O platforms,
> Windows DLL symbol visibility requirements are approximated on these platforms
> by setting default symbol visibility to hidden
> (``-fvisibility-default=hidden``) when building with the following
> configuration:

Reviewer: ```suggestion
Windows DLL symbol visibility requirements are approximated on ELF and MachO platforms
by setting default symbol visibility to hidden
(``-fvisibility-default=hidden``) when building with the following
configuration:
```

::

   LLVM_BUILD_LLVM_DYLIB_VIS=On

For an ELF or Mach-O platform with this setting, the ``LLVM_ABI`` macro is
defined to override the default hidden symbol visibility:

.. code:: cpp

   #define LLVM_ABI __attribute__((visibility("default")))

In addition to ``LLVM_ABI``, there are a few other macros for use in less
common cases described below. All ``LLVM_`` export macros are only for use with
symbols defined in the LLVM library. They must not be used to annotate symbols
defined in other LLVM projects such as lld, lldb, and clang. Their use should

> generally restricted to source code under ``llvm-project/llvm``.

Author: > _if we were to split up LLVM into LLVM.dll and LLVMSupport.dll, we would need two different macros_

How about something like this?

> Export macros are used to annotate symbols only within their intended shared library. This is necessary because of the way Windows handles import/export annotations for shared libraries.

> For example, `LLVM_ABI` resolves to `__declspec(dllexport)` **only** when building source that is part of the LLVM shared library (e.g. source under `llvm-project/llvm`). If `LLVM_ABI` were incorrectly used to annotate a symbol from a different LLVM project—such as Clang—it would always resolve to `__declspec(dllimport)` and the symbol would not be properly exported.

LLVM Interface Export Annotations
=================================
Symbols that are part of LLVM's public interface must be explicitly annotated
to support shared library builds with hidden default symbol visibility. This
document provides background and guidelines for annotating the codebase.

LLVM Shared Library
-------------------
LLVM builds as a static library by default, but it can also be built as a shared
library with the following configuration:

::

   LLVM_BUILD_LLVM_DYLIB=On
   LLVM_LINK_LLVM_DYLIB=On

There are three shared library executable formats we're interested in: PE

> Dynamic Link Library (.DLL) on Windows, Mach-O Shared Object (.dylib) on Apple

Reviewer: ```suggestion
Dynamic Link Library (.dll) on Windows, Mach-O Shared Object (.dylib) on Apple
```

systems, and ELF Shared Object (.so) on Linux, BSD and other Unix-like systems.

ELF and Mach-O Shared Object files can be built with no additional setup or
configuration. This is because all global symbols in the library are exported by

> default-- the same as when building a static library. However, when building a

Reviewer: ```suggestion
default -- the same as when building a static library. However, when building a
```

       return a + b;
     }
   };

.. note::

   When an inline function is annotated for export, the header containing the
   function definition **must** be included by at least one of the library's
   source files or the function will never be compiled with the export
   annotation.

Friend Functions
~~~~~~~~~~~~~~~~
Friend functions declared in a class, struct or union must be annotated with
``LLVM_ABI`` if the corresponding function declaration is also annotated. This
requirement applies even when the class itself is annotated with ``LLVM_ABI``.

.. code:: cpp

   #include "llvm/Support/Compiler.h"

   // An exported function that has friend access to ExampleClass internals.
   LLVM_ABI int friend_function(ExampleClass &obj);

   class ExampleClass {
     // Friend declaration of a function must be annotated the same as the actual
     // function declaration.
     LLVM_ABI friend int friend_function(ExampleClass &obj);
   };

.. note::

   Annotating the friend declaration avoids an “inconsistent dll linkage”
   compiler error when building for Windows. This annotation is harmless but not
   required when building ELF or Mach-O shared libraries.

> VTable and Type Info

Reviewer: ```suggestion
Virtual Table and Type Info
```
